### ISSUE TITLE
--dynamic_local_execution_delay should support individual mnemonics

### ISSUE THREAD

**Issue by DavidANeil** [bazelbuild/bazel#20888]
### Description of the feature request:

We would like to have the ability to customize the delay for different types of actions, while still having a default local delay.

I believe this would be as simple as changing `dynamic_local_execution_delay` to use a `NamedIntegersConverter`.

I believe I have a branch which accomplishes this: #20887. If the proposal is accepted I can finish it up and write tests.

### Which category does this issue belong to?

Local Execution, Remote Execution

**Comment by zhengwei143** [comment]
We actually are hoping to remove `--dynamic_local_execution_delay` eventually, https://github.com/bazelbuild/bazel/pull/20887 brings it in the opposite direction. We are planning to have Blaze make these decisions rather than exposing these toggles through flags. Do you have a compelling reason that requires this functionality per-mnemonic?

**Comment by zhengwei143** [comment]
There is no documentation, and nothing has been implemented yet. The plan to remove `--dynamic_local_execution_delay` stems from it increasing code complexity and likely only benefiting more niche cases. While we happily welcome community contributions, we do have to ensure that they make sense from the maintainability standpoint and also for the broader community.
 
What is your use case for `--dynamic_local_execution_delay` in greater detail? Understanding this would help us better decide the direction of future implementations.

**Comment by DavidANeil** [comment]
We have some expensive actions (execution time ~100s) that occasionally just hang during remote execution, so we currently have these actions as the only actions using dynamic spawn strategy, with a delay of 100s, to help mitigate the remote issue.

We would like to be able to use dynamic spawning globally to better handle cases of spotty internet connections from developer machines, but don't want these large actions to start after only 1s, as they are *very* heavy.

My draft PR would allow us to do this.

**Comment by zhengwei143** [comment]
Is it worth it in this case to use remote execution only for these heavy actions? What are the execution times like locally compared to remote?

Also, I'm not sure whether executing these actions locally after 100s would actually make sense, since its likely that they will probably lose to the remote branch after waiting that long (of course, depending on the local execution times).

**Comment by DavidANeil** [comment]
In our case the remote branch occasionally never finishes (it has hung).

But when it doesn't hang the times are usually comparable (a few seconds slower on remote, unless local is under heavy load, then much faster on remote)

We've only recently enabled remote execution, so we are still trying to work out all of the kinks.

**Comment by zhengwei143** [comment]
How many expensive actions are we talking about here?

What is the main goal here? (I assume there might be some overlap)
1. to ensure that the build completes (hence use dynamic execution in the event that the remote fails) or
2. to not use too much of the user's local machine resources (I assume some element of waiting for 100s weeds out some expensive local actions from being run, which mitigates this to some effect)
3. for better build performance (e.g. end-to-end wall time)

For 1) consider using the fallback strategy running the action on remote and falling back to local if it doesn't work `--strategy=<heavy-mnemonic>=remote,local`.

For 2) consider using `--local_{cpu,ram,extra}_resources` and tune the action's resource set, this will restrict the number of expensive actions running concurrently and prevent it from exploding the user's machine.

For 3) I don't really have a good answer at this point in time given that remote execution is still unreliable. Perhaps trying to make the action less expensive would be a good starting point.

**Comment by DavidANeil** [comment]
We have about 200 of the expensive actions.

1) For the large actions, we mostly just want the build to complete successfully.
We don't have an explicit strategy set, so I believe we have the default `--strategy=remote,worker,sandboxed,local`.

2)
We have tuned the local_resources flags and we attach the `resource_set` (`{"cpu": 6, "memory": 4096}`) to this action appropriately.

3)
This is our `ClosureCompile` action (using https://github.com/google/closure-compiler), so we can't really make it less expensive in terms of CPU time.
We want to enable this flag more generally for better build performance (people with slow internet occasionally spend 60s+ dealing with the remote for actions that take less than 1s to execute locally), we have 4000+ of these actions that can all execute in parallel, so we generally want remote execution, but for the poor connection case we want a sane local experience.

**Comment by zhengwei143** [comment]
Some follow up questions:

1. Remote hanging and slow internet are two separate issues, but just have the same symptom of being a slow remote action correct? 
2. You want to use `--dynamic_local_execution_delay=<heavy-mnemonic>=100s` to weed out the "hanging" remote spawns and run them locally; at this point, there would be fewer local heavy actions to run since a portion of them should have succeeded in the remote. Is this correct?

How about tuning `--experimental_dyamic_slow_remote_time` to 100s? Assuming (1), that could solve this issue more generally across all action mnemonics that have experience a slow remote time rather than for specific mnemonics. All actions run with dynamic execution as normal at first, but once there are local actions still in the queue that have been waiting for >100s for their remote counterpart, they will be prioritized to be run locally next.

**Comment by DavidANeil** [comment]
Correct, they are separate.  I'm not sure they have the same symptom, as in the "hanging" case, the remote action literally never completes.

You are correct on point 2.  In the worst observed case there is only 1 action that ends up completing locally for the heavy mnemonic.

Hmm, I think you are right that `--dynamic_slow_remote_time` might allow us to enable dynamic spawn for all actions.  I'll try it out, and let you know.

**Comment by DavidANeil** [comment]
So, we've been trying to experiment more with this but unfortunately have run into some problems with this when used in our CI by default.

This is the debug log, if anyone happens to know what the solution is here.
```
Caught InterruptedException from ExecutionException for local branch of src/timer.mjs, which may cause a crash.
CancellationException of remote branch of src/timer.mjs, returning null
...
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.AssertionError: Neither branch of src/timer.mjs completed. Local was not cancelled and done and remote was cancelled and done.
	at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.waitBranches(DynamicSpawnStrategy.java:427)
	at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.exec(DynamicSpawnStrategy.java:224)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:261)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1148)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1065)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:562)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
```

**Comment by zhengwei143** [comment]
Do you still have the full INFO logs from that Bazel instance? Also, how often is this happening? Is the local action running as a local spawn or in a worker?

The logs seem to be describing an odd situation:
- local branch is being interrupted: the assumption here that the source of the interruption is either the user or the remote branch seems to be false, since the local future was not canceled. (Correct me if I'm wrong - I'm also assuming here that there weren't any user interrupts since this was running on CI?).
- remote branch is canceled: presumably by the local branch. From the POV of the local branch, it wasn't canceled, so it went ahead and canceled the remote branch. This seems to be working as intended.

So the main question is where this interrupt in the local branch is actually coming from.

**Comment by DavidANeil** [comment]
I do have the full logs, they are 8MB, and contain some file names I can't just post publicly, but if you tell me what parts you need, I can remap the file names for you.

This happens on CI pretty consistently, I haven't been able to get a full CI build to pass since enabling the flag.
I have not had it happen locally.

The local action that is failing is indeed in a worker.  I have noticed locally with `--worker_verbose` that the workers seem to be getting killed quite frequently, so maybe the problem is just that dynamic spawn is interrupting workers instead of doing nothing, or sending a cancellation WorkRequest for `supports-worker-cancellation` actions.
I could imagine a case where an interrupted worker is being given a new WorkRequest while it is still alive during cleanup after the SIGINT, and then failing that subsequent action.

I'll try enabling `experimental_dynamic_ignore_local_signals` to see if I can at least get this case to just allow the remote to continue, though I think there is definitely something smelly going on around workers.

**Comment by zhengwei143** [comment]
Some additional points:
- Which version of Bazel are you currently running on?
- Also, do you set any of `--experimental_total_worker_memory_limit_mb` or `--experimental_worker_memory_limit_mb`? I wonder if workers are hitting actual limits on the machine.
- Side point: It might not be worth it to turn on dynamic execution on CI since their performance benefits are more obvious in incremental builds. That being said, in your case where remote is sometimes unreliable, perhaps its still worth it.
- Any logs around local execution / workers (and why they were killed) should be helpful. Some key terms you could try out (not comprehensive): `WorkerLifecycleManager`, `WorkerSpawnRunner`, `InterruptedException`, `interrupt`, `ExecException`, `WorkerFactory` , `SingleplexWorker`, `SandboxedWorker`.

**Comment by DavidANeil** [comment]
In the last two weeks this has been tested on `7.0.0`, `7.0.1rc1`, `7.0.1`, and `7.0.2`, with no visible difference.

No worker memory limits.

I have been inspecting both the stderr output and the server logs and nothing is standing out, other than the errors I've already posted.

**Comment by zhengwei143** [comment]
> This happens on CI pretty consistently, I haven't been able to get a full CI build to pass since enabling the flag.
I have not had it happen locally.

Has it only been happening after adding `--experimental_dyamic_slow_remote_time`? Or has it been happening before that as well?

Does the action / target (check the `tags`) use `disable-local-prefetch`?

I suspect its a (rare?) race condition that is perhaps exacerbated by your CI environment, considering we haven't heard of similar issues. Are you able to reproduce it on your local machine? What environment is CI running on?

At this point, I don't have enough information to know what is the cause of this. The key should be figuring out the source of the interruption in the local branch (that is not the remote branch canceling it).

**Comment by zhengwei143** [comment]
Any updates with regards to this? Is it still happening?

I think the discussion has strayed from the original feature request - if this is still hapening, I think we should open a separate issue.

**Comment by DavidANeil** [comment]
So we now have dynamic execution running everywhere, but on CI it runs without `worker`: `--dynamic_local_strategy=sandboxed,local`

This works great, though obviously having workers would be great.
Still unsure why it fails so consistently on CI but not locally, as we generally have a pretty consistent setup between the two.

For the purposes of `--dynamic_local_execution_delay`: I no longer need this feature request, so I'll close this issue.
I'll open another issue if I ever get more information on why dynamic execution is killing workers.

**Comment by zhengwei143** [comment]
Additional point: are you using `--experimental_local_lockfree_output` in your CI runners? That could be causing the race
